transport_nantes/mobilito/views.py

```diff
@@ -29,6 +29,111 @@
 logger = logging.getLogger("django")
 
 
+class TutorialState:
+    """Track tutorial pages viewed.
+
+    We track tutorial page views with a short-timeout cookie that
+    represents a set of tutorial pages seen.  If the cookie is absent
+    (and the user is not authenticated with a record of having
+    completed the tutorial), then the set of pages seen is empty.
+    Once the user has seen all the pages, and only then, we permit
+    recording sessions.
+
+    Once the user is authenticated and has completed all pages, we
+    persist that fact to the user record.
+
+    Note that we don't persist tutorial views for authenticated users
+    who have only seen some tutorial records.  They depend on the
+    short timeout cookie.  Once they have seen all pages, we persist
+    the fact that they are done.
+
+    Similarly, unauthenticated users who don't authenticate will time
+    out their tutorial views with the cookie and so have to review the
+    tutorial.
+
+    The tutorial is intended to be short enough that none of this
+    should be particularly bothersome.  Our goal is simply to make
+    sure everyone who records has some chance of knowing what our
+    expectations are, what the different buttons mean.
+
+    """
+    tutorial_cookie_name = 'mobilito_tutorial'
+    all_tutorial_pages = set(['presentation', 'pietons', 'voitures',
+                              'velos', 'transports-collectifs',])
+
+    def default_page(self) -> str:
+        """Return the tutorial page to view if no page requested."""
+        return "presentation"
+
+    def canonical_page(self, page_name) -> str:
+        """Return a valid tutorial page name."""
+        if page_name in self.all_tutorial_pages:
+            return page_name
+        return self.default_page()
+
+    def persist_pages_seen(self, response, pages_seen) -> None:
+        """Persist the tutorial cookie."""
+        response.set_cookie(
+            key=self.tutorial_cookie_name,
+            value=b64encode(pickle.dumps(pages_seen)).hex(),
+            expires=datetime.now(timezone.utc) + timedelta(hours=2))
+
+    def pages_seen(self, request) -> set:
+        """Return the set of pages the user has seen.
+
+        The set is based on the tutorial cookie, and so expires when
+        it does.
+
+        """
+        if 'user' in request and request.user.is_authenticated:
+            mobilito_user = get_MobilitoUser(request)
+            if mobilito_user.completed_tutorial_timestamp:
+                return all_tutorial_pages
```

Comment:
self.all_tutorial_pages
all_tutorial_pages was defined class level and should be accessed through self. 

In this context, evaluating all_tutorial_pages would raise a NameError